4_data_structures/reverse_array.js

```diff
@@ -0,0 +1,44 @@
+/*
+ * Arrays have a reverse method that changes the array by inverting the order in
+ * which its elements appear. For this exercise, write two functions, reverseArray
+ * and reverseArrayInPlace. The first, reverseArray, takes an array as argument
+ * and produces a new array that has the same elements in the inverse order. The
+ * second, reverseArrayInPlace, does what the reverse method does: it modifies
+ * the array given as argument by reversing its elements. Neither may use the
+ * standard reverse method.
+ *
+ * Thinking back to the notes about side effects and pure functions in the
+ * previous chapter, which variant do you expect to be useful in more situations?
+ * Which one runs faster?
+ */
+
+function reverseArray(array) {
+    let reversedArray = [];
+
+    for (let member of array) {
+        reversedArray.unshift(member);
+    }
+
+    return reversedArray;
+}
+
+function reverseArrayInPlace(array) {
+    let reversedArray = [];
+    let countLimit = array.length;
+
+    for (let i = 0; i < countLimit; i++) {
+        let member = array.pop();
+        reversedArray.push(member);
+    }
+    for (let member of reversedArray) {
+        array.push(member);
+    }
+
+    return array;
+}
+
```

Comment:
Entre `reverseArray`, `reverseArrayInPlace` y la función `.reverse` que viene incluída por defecto en JavaScript. ¿Cuál usarías para tu programa?, ¿por qué?, ¿qué diferencias/beneficios tiene usar cada una de las opciones?

Reply:
En general, preferiría usar el `reverseArray`, ya que los otros dos modifican el array in place, y siendo que es mejor practica el hacer copias de un objeto y modificarlas, para evitar comportamientos no deseados y mantener los estados. Por otra parte, me puse a leer sobre el metodo `reverse` y en combinación con el uso de prototype y call, puede aplicarlos a estructuras similares a un array (como un objecto), pero creo que abriría campo de confusión para quienes no esten familiarizados. Me quedo con `reverseArray`.

Reply:
¡Yes! Sí, yo también estoy de acuerdo con que `reverseArray` es la mejor opción. 

En la medida de lo posible, evitemos modificar nuestros objects `{}` o arrays `[]` dado que, por su [naturaleza](https://developer.mozilla.org/en-US/docs/Glossary/Object_reference), cambiar propiedades dentro de ellos puede causar efectos no deseados.